File: 백준/Gold/2665. 미로만들기/미로만들기.java
import java.io.*;
import java.util.*;
public class Main {
	
	static int N, ans;
	static int [][] board, info;
	static final int INF = Integer.MAX_VALUE/2;
	static int [] dx = {1, 0, -1, 0}, dy = {0, 1, 0, -1};
	
	static void solve() {
		ArrayDeque<int[]> q = new ArrayDeque<>();
		board[0][0] = 0;
		q.push(new int[] {0, 0});
		
		while(!q.isEmpty()) {
			int cur[] = q.poll();
			for(int dir = 0 ; dir < 4; dir++) {
				int nx = cur[0] + dx[dir];
				int ny = cur[1] + dy[dir];
				
				if(nx < 0 || nx >= N || ny < 0 || ny >= N) continue;
				
				// 새로운 좌표의 값이 더 크면 작은값으로 갱신
				if(board[nx][ny] > board[cur[0]][cur[1]]) {
					// 검은방인 경우
					if(info[nx][ny] == 0) {
						board[nx][ny] = board[cur[0]][cur[1]]+1;
					}
					else { // 흰 방인경우 
						board[nx][ny] = board[cur[0]][cur[1]];
					}
					
					q.push(new int[] {nx, ny});
				}
			}
		}
	}
	
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		board = new int[N][N];
		info = new int[N][N];
		for(int i = 0; i < N; i++) {
			String temp = br.readLine();
			for(int j = 0; j < temp.length(); j++) {
				board[i][j] = INF;
				info[i][j] = temp.charAt(j) - '0';
			}
		}
		solve();
		System.out.println(board[N-1][N-1] == INF ? 0 : board[N-1][N-1]);
		br.close();
	}
}

File: 백준/Gold/2665. 미로만들기/미로만들기.cc
#include <iostream>
#include <string>
#include <queue>

#define MAX_N 51
#define INF 1e9

using namespace std;

struct Node{
    int x;
    int y;
    int changeNum;

    Node(int _x, int _y, int _changeNum){
        this->x = _x;
        this->y = _y;
        this->changeNum = _changeNum;
    }
};

int N, ans;
int board[MAX_N][MAX_N], vis[MAX_N][MAX_N];
int dx[4] = {1, 0, -1, 0}, dy[4] = {0, 1, 0, -1};



int solve(){
    queue<Node> q;
    vis[0][0] = 0;
    q.push(Node(0, 0, 0));

    while(!q.empty()){
        Node cur = q.front();
        q.pop();
        for(int dir = 0; dir < 4; dir++){
            int nx = cur.x + dx[dir];
            int ny = cur.y + dy[dir];
            int changeNum = cur.changeNum;
            
            if(nx < 0 || nx >= N || ny < 0 || ny >= N) continue;
            // 변경 횟수가 클 경우 방문할 필요 x
            if(vis[nx][ny] != INF && vis[nx][ny] <= changeNum) continue;
            vis[nx][ny] = changeNum;
            // 검은방인 경우
            if(board[nx][ny] == 0) q.push(Node(nx, ny, changeNum+1));
            else q.push(Node(nx, ny, changeNum));
        }
    }

    return vis[N-1][N-1];
}

int main() {
    cin.tie(0);
    ios::sync_with_stdio(0);
    cin >> N;
    for(int i = 0; i < N; i++){
        string temp; cin >> temp;
        for(int j = 0; j < temp.length(); j++){
            board[i][j] = temp[j] - '0'; // 방 정보
            vis[i][j] = INF; // 방문횟수 초기화
        }
    }
    
    cout << solve();
    return 0;
}